File: src/main/java/fr/deroffal/aoc/Day02.java
package fr.deroffal.aoc;

import java.util.Collections;
import java.util.List;

import fr.deroffal.aoc.util.Utils;

public class Day02 {

	public static void main(String[] args) {
		List<String> input = Utils.readLines("day02.txt");

		System.out.println("Partie 1 : " + calculerDifferenceExtremum(input));
		System.out.println("Partie 2 : " + calculerResteDivision(input));

	}

	private static int calculerDifferenceExtremum(final List<String> input) {
		return input.stream().mapToInt(Day02::calculerDifference).sum();
	}

	static int calculerDifference(final String s) {
		final List<Integer> liste = Utils.splitToIntegerList("\t", s);
		return Collections.max(liste) - Collections.min(liste);
	}

	private static int calculerResteDivision(final List<String> input) {
		return input.stream().mapToInt(Day02::calculerDifferenceDivision).sum();
	}

	static int calculerDifferenceDivision(final String s) {
		final List<Integer> liste = Utils.splitToIntegerList("\t", s);
		liste.sort(Collections.reverseOrder());

		int listeSize = liste.size();
		for (int x = 0; x < listeSize; x++) {
			int xValue = liste.get(x);
			for (int y = x + 1; y < listeSize; y++) {
				int yValue = liste.get(y);
				if (xValue % yValue == 0) {
					return xValue / yValue;
				}
			}
		}
		throw new IllegalArgumentException("Un résultat aurait dû être trouvé!");
	}

}

File: kotlin/src/main/kotlin/fr/deroffal/aoc/kotlin/Day02.kt
package fr.deroffal.aoc.kotlin

import fr.deroffal.aoc.kotlin.tools.readLines

fun main(args: Array<String>) {
    val input = readLines("day02.txt")
    val day02 = Day02()
    println("Part 1 : ${day02.solvePart1(input)}")
    println("Part 2 : ${day02.solvePart2(input)}")
}

class Day02 {
    internal fun solvePart1(input: List<String>): Int =
            parseInputToIntList(input)                  //parse to list of int list
                    .map { it.sorted() }                //sort each list
                    .map { it.last() - it.first() }     //subtract min to max
                    .sum()                              //add

    internal fun solvePart2(input: List<String>): Int =
            parseInputToIntList(input)
                    .map { findCandidates(it) }
                    .sumBy { it.first / it.second }

    private fun parseInputToIntList(input: List<String>) = input.map { it.split(" ").map { it.toInt() } }

    private fun findCandidates(row: List<Int>): Pair<Int, Int> =
            row.flatMap { int -> row.map { Pair(int, it) } }    //each number with each of the line
                    .filter { it.first != it.second }               //not twice the same
                    .first { it.first % it.second == 0 }            //find the firsts matching
}